design/external-references.md

```diff
@@ -0,0 +1,344 @@
+- **REQ3**: Allow CDK apps to reference resources that were created elsewhere (by other CDK apps, manually or any other mean) by referencing the physical identity. For example, an S3 Bucket ARN should be sufficient in order to reference
+an existing S3 bucket.
+- **REQ4**: If possible, implement the mechanism such that it can be later used to publish and consume constructs through other key/value mechanism such as environment variables, SSM parameters, etc.
+- **REQ5**: If a resource is exported twice under the same export name, only a single set of outputs will be created ([#1496](https://github.com/awslabs/aws-cdk/issues/1496)).
+- **REQ6**: If the same resource is imported twice in the same stack, the "import" method will return the same object.
+- **REQ7**: For some resources (like VPC for example), it should be possible to import the resource by retrieving it's details from the deployed environment via a context provider.
+
+## Approach
+
+As usual, we'll take a layered approach.
+
+**Serializable constructs**: At the low-level, we'll define what it means for a construct to be "__serializable__" through string key/value context. The mechanism will be recursive such that it will be possible to write/read singular values and write/read serializable objects (which, themselves will write/read their own values and so forth).
+
+**Serialization context**: the notion of a "string key/value context" represents the lowest common denominator for serialization, and specifically is the only type supported by CloudFormation's cross-stack export/import mechanism. If a construct can serialize itself through a set of string key/values, we can pass it via CloudFormation's import/export, environment variables, SSM parameters, etc.
+
+**CloudFormation imports/exports as a serialization context**: on top of that, we will implement a serialization/deserialization context based on AWS CloudFormation's import/exports mechanism. "writing" a value means "output/export" and "reading" a value means "import". The "export name" will be used as a prefix that represent an object, and sub-objects will be serialized by adding another component to the export name.
+
+**Convenience methods**: now that constructs can be serializable through imports/exports, we can implement a set of convenience methods for each AWS resource to provide nice ergonomics for the specific use case of exporting and importing resources across apps.
+
+## Design
+
+### Serialization
+
+If a construct is said to be serializable, it must implement the `ISerializable` interface:
+
+```ts
+interface ISerializable {
+  serialize(ctx: ISerializationContext): void;
+}
+```
+
+When an object is serialized, the `serialize` method is called with an object which implements the following interface:
+
+```ts
+interface ISerializationContext {
+  writeString(key: string, value: string, options?: SerializationOptions): void
```

Comment:
We're going to need `ListOfStrings` as well.

Reply:
I am not sure there’s a safe generic way to implement that. Maybe we can json-serialize?

Reply:
That's not very useful: they won't be *usable* at deployment time, and they won't be *available* at synthesis time.

I believe CFN will allow us to export/import a list-of-strings. That's fine. We just have to make a decision on whether this mechanism is going to work deployment-time or synthesis-time.

Reply:
I’ll try to see if we can export a list through CFN outputs.

Reply:
Well:

```
The Value field of every Outputs member must evaluate to a String.
```

Reply:
Bollocks. I truly thought that was possible.

Reply:
Well we can export a `{Fn::Join}` and import with a `{Fn::Split}`...

Reply:
I think you'll love the changes I am about to post to the RFC 😉

Reply:
The `Fn::Join`/`Fn::Split` requires you're able to admit a delimiter that is guaranteed to not be present in the values.